Centralize Loading code for Shell/Pages

michael-hawker opened the pull request:
## Summary of the Pull Request

This PR is ready to review with a bunch of code clean-up around shell and page initialization/navigation, progress, errors, and the back button.

There's still some future items to think about, but that's beyond the scope of this work right now. See more notes in comments below on changes:

> Alright, I cleaned up all the Page Initialization and Navigation code for Shell, it's all centralized there now (which makes sense) and removes 3x duplicated code we had in the ExtPage templates. They're just basic XAML controls now which don't really need to worry about how they get loaded, and just get their VMs.
>
> I also tidied up the massive if statement we had in ShellPage.xaml.cs for each page type so we have a singular logical flow that just branches in two statements with an inline switch for the VM constructor and the page type.
> 
> In the future, we may want to cache some page VMs there, so if you navigate back/forth we don't reload everything...
>
> This also let us remove a message which was effectively just internal communication. Should be a bit cleaner now, certainly easier to add new page types now too.

These are some of the last remnants of things from the original MVVM setup when we were still testing patterns and then just copied.

This uses `IsInitialized` for tracking the progress bar and bubbles up errors better without blocking the app (and provides better context for errors).

This also fixes the issue with the Back Button stealing focus instead of the search box (by moving it to the Shell page to leave the SearchBox control as just responsible for the searchbox/keyboard input).

-------

Future Open questions/TODO:
- [ ] How do we better track the parent extension/owner of pages/commands/etc... for error reporting?
- [x] Need to clean-up backend logic of VMs for the loading state for the new Shell page centralization
  - [x] For instance, progress only happens like once now...
  - [x] May be able to centralize some of the loading bits logic?
- [x] Finalize some brushes for error items
- [ ] Still need to potentially track individual issues loading specific items, but maybe this is a good enough start for a single PR once the above are addressed?
- [ ] Cache page VMs?
- [ ] Clean-up async loading of pages and their content

Helps with #220.
This will make #238 easier as then we don't need as much XAML/Loading glue for those

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [ ] **Closes:** #xxx
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

niels9001 commented on src/modules/cmdpal/Microsoft.CmdPal.UI/ShellPage.xaml:
```diff
@@ -105,23 +96,58 @@
+                MaxHeight="120"
+                Background="{ThemeResource SystemControlErrorBackgroundColor}"
+                BorderBrush="{ThemeResource SystemControlErrorTextForegroundBrush}"
+                BorderThickness="0,1,0,0"
```
```suggestion
CornerRadius="0"
                BorderThickness="0,1,0,1"
```

michael-hawker commented:
Alright, the `IsLoading` reporting by the extension seems a bit wonky. Seems to start as `false` and some of them don't report it like the main page does, so that's messing with this a bit compared to the logic we had before. Needs some more investigation.

michael-hawker commented on src/modules/cmdpal/Microsoft.CmdPal.UI/ExtViews/ListPage.xaml.cs:
```diff
@@ -52,94 +39,29 @@ public ListPage()
-            }
-        }
-
-        if (e.NavigationMode == NavigationMode.Back)
-        {
-            // Upon navigating _back_ to this page, immediately select the
-            // first item in the list
-            ItemsList.SelectedIndex = 0;
```
Ah, there's some bad merge to clean-up here. Taking a look.

michael-hawker commented:
Alright, I think I've resolved the merge conflicts with main.

However, there's two things (which I think are related) which are regressions, so need to be fixed:
- [ ] Back Button doesn't update to Search icon when back on home page
- [ ] Home page won't filter again after first launch...

michael-hawker commented:
Yup, pretty sure the two issues above are from the main view model for the main list page not getting reset when we're going back.

Tracked it down comparing the old behavior off `main`: 

1) In ShellPage.xaml.cs:GoBack we navigate back
2) Then this triggers OnNavigatedTo again in ListPage.xaml.cs
3) This called the `NavigateToPageMessage` again on itself (which we removed this message in this PR)
4) This of course, updated the Shell's CurrentPage VM

So, the question is what's the proper fix to resolve this in the new setup. I either need to bring back the message as we need to track the navigations OR hopefully we could do something in the GoBack method, but that may mean tracking the state... not sure. Going to investigate options now.

At least the good news, is once the state is properly tracked, then we should be good to go and the back button and the filtering should just work again.

This may be a general follow-up question is how much we want to rely on the Frame and its navigation stack to maintain states vs. managing that ourselves with a stack in our ShellViewModel. Will there be cases where we want to shortcut the back stack between extensions or commands somehow?

michael-hawker commented on src/modules/cmdpal/Microsoft.CmdPal.UI/ExtViews/ListPage.xaml.cs:
```diff
@@ -35,39 +34,16 @@ public ListViewModel? ViewModel
-            else
-            {
-                ViewModel = lvm;
-                WeakReferenceMessenger.Default.Send<NavigateToPageMessage>(new(lvm));
```
Just for future reference, this is the missing call that's causing us to miss the change of the current page VM when we GoBack.

michael-hawker replied:
This is now fixed by tracking all navigations of the root frame in the ShellPage.xaml.cs and using that to get the parameter and track the CurrentPage VM centrally there itself.

michael-hawker commented on src/modules/cmdpal/Microsoft.CmdPal.UI/ExtViews/ListPage.xaml.cs:
```diff
@@ -86,49 +62,6 @@ protected override void OnNavigatedTo(NavigationEventArgs e)
-    LoadedState = result ? ViewModelLoadedState.Loaded : ViewModelLoadedState.Error;
-
-    // Immediately select the first item in the list
-    ItemsList.SelectedIndex = 0;
```
This is the only difference in the initialization pattern here from the central one, but i think this is redundant as we have the check in the ItemsUpdated event to do this once the page is loaded...

zadjii commented on src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/Microsoft.CmdPal.UI.ViewModels.csproj:
```diff
@@ -13,6 +13,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+	<PackageReference Include="CommunityToolkit.Common" />
```
nit tab

zadjii reviewed (approved):
Okay yea I only have nits
